Source/WebCore/dom/Document.h

```diff
@@ -441,6 +441,9 @@ class Document
 
     bool canContainRangeEndPoint() const final { return true; }
 
+    bool shouldFireMutationEvents() const { return m_shouldFireMutationEvents; }
```

Comment:
It seems unfortunate that we have to create this new mechanism, particularly a global one for the entire document. Is there some way we can do this with something we already have? There are already so many ways we do things without firing mutation events, such as the "parser" mode for adding children. Is this really the first time we had to manipulate a document without firing mutation events?

Reply:
> Is this really the first time we had to manipulate a document without firing mutation events?

I had wondered about the same thing — and checked before writing up the patch, but didn’t find anything.

And it did feel some hesitation about needing to add it globally — but at https://html.spec.whatwg.org/multipage/dom.html#concept-document-fire-mutation-events-flag that spec says:

> Each [`Document`](https://html.spec.whatwg.org/multipage/dom.html#document) has a fire mutation events flag, which is a boolean, initially true.

That is, the spec explicitly says it’s a flag on `Document`. And the existing WPT tests do include cases for checking that the effect is global.

Still, I’d hoped a clever alternative for implementing it without hanging more onto `Document` might emerge during review. I messed around with trying some things locally, but I couldn’t make anything else work.